File: traversal/src/main/scala/overflowdb/algorithm/GnnExporter.java
package overflowdb.algorithm;

import overflowdb.Edge;
import overflowdb.Node;
import gnu.trove.list.array.TLongArrayList;

import java.util.*;

/**
 * Exporter utility to extract graph structures into flat primitive arrays,
 * optimized for consumption by Graph Neural Network (GNN) frameworks.
 */
public class GnnExporter {

    public static class GnnExport {
        public final long[] nodeIds;
        public final String[] nodeLabels;
        public final long[] edgeSrcIds;
        public final long[] edgeDstIds;
        public final String[] edgeLabels;

        public GnnExport(long[] nodeIds, String[] nodeLabels, long[] edgeSrcIds, long[] edgeDstIds, String[] edgeLabels) {
            this.nodeIds = nodeIds;
            this.nodeLabels = nodeLabels;
            this.edgeSrcIds = edgeSrcIds;
            this.edgeDstIds = edgeDstIds;
            this.edgeLabels = edgeLabels;
        }
    }

    /**
     * Exports the subgraph induced by the given collection of nodes into a flat primitive representation.
     * @param nodes Collection of nodes forming the induced subgraph.
     * @return A GnnExport instance containing parallel primitive arrays.
     */
    public static GnnExport exportGraph(Collection<Node> nodes) {
        int nodeCount = nodes.size();
        long[] nodeIds = new long[nodeCount];
        String[] nodeLabels = new String[nodeCount];
        
        Set<Long> nodeSet = new HashSet<>(nodeCount);
        int idx = 0;
        for (Node node : nodes) {
            long id = node.id();
            nodeIds[idx] = id;
            nodeLabels[idx] = node.label();
            nodeSet.add(id);
            idx++;
        }

        TLongArrayList srcList = new TLongArrayList();
        TLongArrayList dstList = new TLongArrayList();
        List<String> labelList = new ArrayList<>();

        for (Node node : nodes) {
            Iterator<Edge> outEdges = node.outE();
            while (outEdges.hasNext()) {
                Edge edge = outEdges.next();
                Node inNode = edge.inNode();
                long inId = inNode.id();
                if (nodeSet.contains(inId)) {
                    srcList.add(node.id());
                    dstList.add(inId);
                    labelList.add(edge.label());
                }
            }
        }

        return new GnnExport(
            nodeIds,
            nodeLabels,
            srcList.toArray(),
            dstList.toArray(),
            labelList.toArray(new String[0])
        );
    }
}

File: formats/src/main/scala/overflowdb/formats/gnn/GnnExporter.scala
package overflowdb.formats.gnn

import overflowdb.algorithm.GnnExporter as GnnArrays
import overflowdb.formats.{ExportResult, Exporter, resolveOutputFileSingle}
import overflowdb.{Edge, Graph, Node}

import java.io.BufferedWriter
import java.nio.file.{Files, Path}
import scala.jdk.CollectionConverters.IteratorHasAsScala
import scala.util.Using

/** Writes the graph as a small set of parallel arrays in JSON, ready to be loaded by a Graph Neural
  * Network pipeline. The layout mirrors the in-memory representation produced by
  * [[overflowdb.algorithm.GnnExporter]]: node ids and labels, then the source id, destination id
  * and label of every edge that stays within the exported node set.
  *
  * Only edges between exported nodes are emitted, so passing a slice or any other induced subgraph
  * produces a self-contained file with no dangling endpoints.
  */
object GnnExporter extends Exporter:
    override def defaultFileExtension = "json"

    override def runExport(graph: Graph, outputFile: Path): ExportResult =
        runExport(graph.nodes().asScala, Iterator.empty, outputFile)

    override def runExport(
      nodes: IterableOnce[Node],
      edges: IterableOnce[Edge],
      outputFile: Path
    ): ExportResult =
        val nodeList = new java.util.ArrayList[Node]()
        nodes.iterator.foreach(nodeList.add)
        val arrays  = GnnArrays.exportGraph(nodeList)
        val outFile = resolveOutputFileSingle(outputFile, s"export.$defaultFileExtension")

        Using.resource(Files.newBufferedWriter(outFile)) { writer =>
            writer.write("{")
            writeLongArray(writer, "nodeIds", arrays.nodeIds)
            writer.write(",")
            writeStringArray(writer, "nodeLabels", arrays.nodeLabels)
            writer.write(",")
            writeLongArray(writer, "edgeSrcIds", arrays.edgeSrcIds)
            writer.write(",")
            writeLongArray(writer, "edgeDstIds", arrays.edgeDstIds)
            writer.write(",")
            writeStringArray(writer, "edgeLabels", arrays.edgeLabels)
            writer.write("}")
            writer.newLine()
        }

        ExportResult(
          nodeCount = arrays.nodeIds.length,
          edgeCount = arrays.edgeSrcIds.length,
          files = Seq(outFile),
          additionalInfo = None
        )
    end runExport

    private def writeLongArray(writer: BufferedWriter, name: String, values: Array[Long]): Unit =
        writer.write(s"\"$name\":[")
        var i = 0
        while i < values.length do
            if i > 0 then writer.write(",")
            writer.write(values(i).toString)
            i += 1
        writer.write("]")

    private def writeStringArray(
      writer: BufferedWriter,
      name: String,
      values: Array[String]
    ): Unit =
        writer.write(s"\"$name\":[")
        var i = 0
        while i < values.length do
            if i > 0 then writer.write(",")
            writer.write(escape(values(i)))
            i += 1
        writer.write("]")

    private def escape(value: String): String =
        val sb = new StringBuilder(value.length + 2)
        sb.append('"')
        var i = 0
        while i < value.length do
            value.charAt(i) match
                case '"'  => sb.append("\\\"")
                case '\\' => sb.append("\\\\")
                case '\n' => sb.append("\\n")
                case '\r' => sb.append("\\r")
                case '\t' => sb.append("\\t")
                case c    => sb.append(c)
            i += 1
        sb.append('"')
        sb.toString
end GnnExporter
